README
Open-BLDC high speed register based debug and control protocol library.


NEWS
No major news to report...


tests/check_ring_suite.c
/*
 * libgovernor - Open-BLDC configuration and debug protocol library
 * Copyright (C) 2010 by Piotr Esden-Tempski <piotr@esden.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <check.h>

#include <string.h>

#include "lg/ring.h"

#include "check_suites.h"

struct ring test_ring;
uint8_t data[10];

void init_ring_tc(void)
{
	ring_init(&test_ring, data, 10);
}

void clean_ring_tc(void)
{
}


START_TEST(test_ring_create)
{
	fail_unless(9 == RING_SIZE(&test_ring));
	fail_unless(data == RING_DATA(&test_ring));
}
END_TEST

START_TEST(test_ring_write_read_one)
{
	int32_t ret;
	uint8_t ch;

	ret = ring_write_ch(&test_ring, (uint8_t)'A');

	fail_unless('A' == ret);

	ret = ring_read_ch(&test_ring, &ch);

	fail_unless('A' == ret);
	fail_unless('A' == ch);
}
END_TEST

START_TEST(test_ring_write_read_max)
{
	int32_t ret;
	uint8_t i, ch;

	for(i=0; i<9; i++){
		ret = (uint8_t)ring_write_ch(&test_ring, i);

		fail_unless(i == ret);
	}

	fail_unless(-1 == ring_write_ch(&test_ring, 10));

	for(i=0; i<9; i++){
		ret = (uint8_t)ring_read_ch(&test_ring, &ch);

		fail_unless(i == ret);
		fail_unless(i == ch);
	}

	ch = 10;
	fail_unless(-1 == ring_read_ch(&test_ring, &ch));
	fail_unless(10 == ch);
}
END_TEST

START_TEST(test_ring_write_read_array)
{
	uint8_t array[10];
	uint8_t array1[] = "ABCD";
	uint8_t array2[] = "EFGHIJ";

	fail_unless(4 == ring_write(&test_ring, array1, 4));
	fail_unless(-5 == ring_write(&test_ring, array2, 6));

	memset(array, 0, 10);
	fail_unless(9 == ring_read(&test_ring, array, 10));
	fail_unless(0 == memcmp(array, "ABCDEFGHI", 9));
	fail_unless(0 != memcmp(array, "ABCDEFGHIJ", 10));

	memset(array, 0, 10);
	fail_unless(4 == ring_write(&test_ring, (uint8_t *)"ABCD", 4));
	fail_unless(4 == ring_read(&test_ring, array, 10));
	fail_unless(0 == memcmp(array, "ABCD", 4));

	memset(array, 0, 10);
	fail_unless(6 == ring_write(&test_ring, (uint8_t *)"ABCDEF", 6));
	fail_unless(-4 == ring_read(&test_ring, array, 4));
	fail_unless(0 == memcmp(array, "ABCD", 4));
}
END_TEST

Suite *make_lg_ring_suite()
{
	Suite *s;
	TCase *tc_ring_read_write;

	s= suite_create("Ring buffer");

	tc_ring_read_write = tcase_create("Buffer read/write");
	suite_add_tcase(s, tc_ring_read_write);
	tcase_add_checked_fixture(tc_ring_read_write, init_ring_tc, clean_ring_tc);
	tcase_add_test(tc_ring_read_write, test_ring_create);
	tcase_add_test(tc_ring_read_write, test_ring_write_read_one);
	tcase_add_test(tc_ring_read_write, test_ring_write_read_max);
	tcase_add_test(tc_ring_read_write, test_ring_write_read_array);

	return s;
}


tests/check_gprot_suite.c
/*
 * libgovernor - Open-BLDC configuration and debug protocol library
 * Copyright (C) 2010 by Piotr Esden-Tempski <piotr@esden.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <check.h>

#include <string.h>
#include <stdio.h>

#include "lg/gpdef.h"
#include "lg/gprotm.h"
#include "lg/gprotc.h"

#include "check_utils.h"
#include "check_suites.h"

uint16_t gp_register_map[32];

int gpm_register_changed = 0;
int gpm_register_changed_addr = 0;

int gpm_string_received = 0;
char gpm_string_received_string[1024];
int gpm_string_received_len = 0;
int gpc_get_version = 0;

void gpm_trigger_output_hook(void *data)
{
        int32_t dat;

	data = data;

	while(-1 != (dat = gpm_pickup_byte())){
		if(gpc_handle_byte(dat))
			printf("gpc_handle_byte %02X failed.\n", dat);
	}
}

void gpm_register_changed_hook(void *data, uint8_t addr)
{
	data = data;

	gpm_register_changed_addr = addr;
	gpm_register_changed = 1;
}

void gpc_trigger_output_hook(void* data)
{
	int32_t dat;

	data = data;

	while(-1 != (dat = gpc_pickup_byte())){
		if(gpm_handle_byte(dat))
			printf("gpm_handle_byte %02X failed.\n", dat);
	}
}

void gpm_string_received_hook(void *data, char *string, int len)
{
	data = data;

	gpm_string_received = 1;
	memcpy(gpm_string_received_string+gpm_string_received_len, string, len);
	gpm_string_received_len += len;
}

void gpc_get_version_hook(void *data)
{
	data = data;

	gpc_get_version = 1;
}

void init_gprot_tc(void)
{
	int i;

	gpm_init(gpm_trigger_output_hook, NULL, gpm_register_changed_hook, NULL);
	gpm_set_string_received_callback(gpm_string_received_hook, NULL);

	gpc_init(gpc_trigger_output_hook, NULL, NULL, NULL);
	gpc_set_get_version_callback(gpc_get_version_hook, NULL);

	for(i=0; i<32; i++){
		gp_register_map[i] = 0xAA55+i;
		gpc_setup_reg(i, &gp_register_map[i]);
	}
}

void clean_gprot_tc(void)
{
	gpm_register_changed = 0;
	gpm_register_changed_addr = 0;
	gpm_string_received = 0;
	memset(gpm_string_received_string, 0, sizeof(gpm_string_received_string));
	gpm_string_received_len = 0;
}

START_TEST(test_gprot_write)
{
	uint8_t addr = 0;

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpm_send_set(addr, 0x0000));
		fail_unless(0x0000 == gp_register_map[addr]);
		fail_unless(0x0000 == gpm_get_register_map_val(addr));
	}

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpm_send_set(addr, 0xAA55));
		fail_unless(0xAA55 == gp_register_map[addr]);
		fail_unless(0xAA55 == gpm_get_register_map_val(addr));
	}

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpm_send_set(addr, 0xFFFF));
		fail_unless(0xFFFF == gp_register_map[addr]);
		fail_unless(0xFFFF == gpm_get_register_map_val(addr));
	}
}
END_TEST

START_TEST(test_gprot_read)
{
	uint8_t addr;

	for(addr=0; addr<32; addr++){
		gp_register_map[addr] = 0x0000;
	}

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpm_send_get(addr));
		fail_unless(0x0000 == gpm_get_register_map_val(addr));
	}

	for(addr=0; addr<32; addr++){
		gp_register_map[addr] = 0xAA55;
	}

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpm_send_get(addr));
		fail_unless(0xAA55 == gpm_get_register_map_val(addr));
	}

	for(addr=0; addr<32; addr++){
		gp_register_map[addr] = 0xFFFF;
	}

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpm_send_get(addr));
		fail_unless(0xFFFF == gpm_get_register_map_val(addr));
	}
}
END_TEST

START_TEST(test_gprot_read_write)
{
	uint8_t addr;

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpm_send_set(addr, 0x0000+addr));
		fail_unless(0 == gpm_send_get(addr));
		fail_unless(0x0000+addr == gpm_get_register_map_val(addr));
	}

}
END_TEST

START_TEST(test_gprot_send_short_string)
{
	int i;
	char *string = "Hello World!";
	int string_size = strlen(string);

	fail_unless(string_size == gpc_send_string(string, string_size));

	fail_unless(string_size == gpm_string_received_len);
	fail_unless(1 == gpm_string_received);

	for (i=0; i<string_size; i++) {
		fail_unless(string[i] == gpm_string_received_string[i]);
	}

	fail_unless(0 == gpm_string_received_string[i]);
}
END_TEST

START_TEST(test_gprot_send_long_string)
{
	int i;
	int string_size = (('z' - 'a') + 1) * ((('9' - '0') + 1) + 1);
	char string[string_size];

	/* Generate string. */
	{
		int gi, gj, gk = 0;

		for(gi=0; gi<(('z' - 'a') + 1); gi++) {
			string[gk++] = 'a'+gi;
			for(gj=0; gj<(('9' - '0') + 1); gj++){
				string[gk++] = '0'+gj;
			}
		}

		string[gk] = '\0';

	}

	fail_unless(string_size == gpc_send_string(string, string_size));

	fail_unless(string_size == gpm_string_received_len);
	fail_unless(1 == gpm_string_received);

	for (i=0; i<string_size; i++) {
		fail_unless(string[i] == gpm_string_received_string[i]);
	}

	fail_unless(0 == gpm_string_received_string[i]);
}
END_TEST

START_TEST(test_gprot_send_arbitrary_string)
{
	int i, j;
	int string_size = (('z' - 'a') + 1) * ((('9' - '0') + 1) + 1);
	char string[string_size];

	/* Generate string. */
	{
		int gi, gj, gk = 0;

		for(gi=0; gi<(('z' - 'a') + 1); gi++) {
			string[gk++] = 'a'+gi;
			for(gj=0; gj<(('9' - '0') + 1); gj++){
				string[gk++] = '0'+gj;
			}
		}

		string[gk] = '\0';

	}

	for (j=0; j<=string_size; j++) {

		fail_unless(j == gpc_send_string(string, j));

		fail_unless(j == gpm_string_received_len);
		fail_unless(1 == gpm_string_received);

		for (i=0; i<j; i++) {
			fail_unless(string[i] == gpm_string_received_string[i]);
		}

		fail_unless(0 == gpm_string_received_string[i]);

		gpm_string_received_len = 0;
		gpm_string_received = 0;
		memset(gpm_string_received_string, 0, sizeof(gpm_string_received_string));
	}

}
END_TEST

START_TEST(test_gprot_get_client_version)
{
	fail_unless(0 == gpm_send_get_version());

	fail_unless(1 == gpc_get_version);

	fail_unless(0 == regmatch("^libgovernor [[:digit:]]+\\.[[:digit:]]+-[[:alnum:]]{8}(-dirty)?, build [[:digit:]]{8}$", gpm_string_received_string));
	fail_unless(0 == regmatch("^Copyright \\(C\\) 2010-20[[:digit:]]{2} Piotr Esden-Tempski <piotr@esden.net>$", gpm_string_received_string));
	fail_unless(0 == regmatch("^License GPLv3\\+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>$", gpm_string_received_string));

	memset(gpm_string_received_string, 0, sizeof(gpm_string_received_string));
}
END_TEST

Suite *make_lg_gprot_suite()
{
	Suite *s;
	TCase *tc;

	s= suite_create("Governor protocol master and client");

	tc = tcase_create("Basic functions");
	suite_add_tcase(s, tc);
	tcase_add_checked_fixture(tc, init_gprot_tc, clean_gprot_tc);
	tcase_add_test(tc, test_gprot_write);
	tcase_add_test(tc, test_gprot_read);
	tcase_add_test(tc, test_gprot_read_write);
	tcase_add_test(tc, test_gprot_send_short_string);
	tcase_add_test(tc, test_gprot_send_long_string);
	tcase_add_test(tc, test_gprot_send_arbitrary_string);
	tcase_add_test(tc, test_gprot_get_client_version);

	return s;
}


tests/check_gprotc_suite.c
/*
 * libgovernor - Open-BLDC configuration and debug protocol library
 * Copyright (C) 2010-2011 by Piotr Esden-Tempski <piotr@esden.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <check.h>

#include <string.h>
#include <stdio.h>

#include "lg/gpdef.h"
#include "lg/gprotc.h"

#include "check_utils.h"
#include "check_suites.h"

uint16_t gpc_dummy_register_map[32];

void *gpc_dummy_trigger_output_data = 0;
int gpc_dummy_trigger_output_triggered = 0;
void *gpc_dummy_register_changed_data = 0;
int gpc_dummy_register_changed = 0;
int gpc_dummy_register_changed_addr = 0;
void *gpc_dummy_get_version_data = 0;
int gpc_dummy_get_version_triggered = 0;

void gpc_dummy_trigger_output_hook(void* data)
{
	gpc_dummy_trigger_output_data = data;
	gpc_dummy_trigger_output_triggered = 1;
}

void gpc_dummy_register_changed_hook(void* data, uint8_t addr)
{
	gpc_dummy_register_changed_data = data;
	gpc_dummy_register_changed = 1;
	gpc_dummy_register_changed_addr = addr;
}

void gpc_dummy_get_version_hook(void* data)
{
	gpc_dummy_get_version_data = data;
	gpc_dummy_get_version_triggered = 1;
}

void init_gprotc_tc(void)
{
	int i;

	for(i=0; i<32; i++)
		gpc_dummy_register_map[i] = 0xAA55+i;

	gpc_init(gpc_dummy_trigger_output_hook, (void *)1, gpc_dummy_register_changed_hook, (void*)1);
	gpc_set_get_version_callback(gpc_dummy_get_version_hook, (void *)1);

	gpc_dummy_trigger_output_data = 0;
	gpc_dummy_trigger_output_triggered = 0;
	gpc_dummy_register_changed_data = 0;
	gpc_dummy_register_changed = 0;
	gpc_dummy_register_changed_addr = 0;
	gpc_dummy_get_version_data = 0;
	gpc_dummy_get_version_triggered = 0;
}

void clean_gprotc_tc(void)
{
}

START_TEST(test_gprotc_send_reg)
{
	int i;
	uint8_t addr;

	addr = 0;
	for(i = 0; i<256; i++){
		fail_unless(1 == gpc_send_reg(addr));
		fail_unless(-1 == gpc_pickup_byte());
		addr++;
	}

	fail_unless(0 == gpc_setup_reg(0, &gpc_dummy_register_map[0]));
	fail_unless(0 == gpc_send_reg(0));
	fail_unless(0 == gpc_pickup_byte());
	fail_unless(0x55 == gpc_pickup_byte());
	fail_unless(0xAA == gpc_pickup_byte());
	fail_unless(-1 == gpc_pickup_byte());
	fail_unless(1 == gpc_send_reg(1));

	for(addr=1; addr<32; addr++){
		fail_unless(0 == gpc_setup_reg(addr, &gpc_dummy_register_map[addr]));
		fail_unless(0 == gpc_send_reg(addr));
		fail_unless(addr == gpc_pickup_byte());
		fail_unless(0x55+addr == gpc_pickup_byte());
		fail_unless(0xAA == gpc_pickup_byte());
		fail_unless(-1 == gpc_pickup_byte());
	}


	for(addr=32; addr!=0; addr++){
		fail_unless(1 == gpc_send_reg(addr));
		fail_unless(-1 == gpc_pickup_byte());
	}
}
END_TEST

START_TEST(test_gprotc_handle_byte_read)
{
	uint8_t addr = 0;

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpc_setup_reg(addr, &gpc_dummy_register_map[addr]));
	}

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpc_handle_byte(addr | GP_MODE_READ | GP_MODE_PEEK));
		fail_unless(1 == gpc_dummy_trigger_output_triggered);
		fail_unless((void *)1 == gpc_dummy_trigger_output_data);
		fail_unless(addr == gpc_pickup_byte());
		fail_unless(0x55+addr == gpc_pickup_byte());
		fail_unless(0xAA == gpc_pickup_byte());
		fail_unless(-1 == gpc_pickup_byte());

		gpc_dummy_trigger_output_triggered = 0;
		gpc_dummy_trigger_output_data = 0;
	}

	for(addr=0; addr<32; addr++){
		fail_unless(1 == gpc_handle_byte(addr | GP_MODE_READ | GP_MODE_PEEK | GP_MODE_STRING));
		fail_unless(0 == gpc_dummy_trigger_output_triggered);
		fail_unless((void *)0 == gpc_dummy_trigger_output_data);
		fail_unless(-1 == gpc_pickup_byte());
	}
}
END_TEST

START_TEST(test_gprotc_handle_byte_write)
{
	uint8_t addr = 0;
	uint16_t data = 0xDADE;

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpc_setup_reg(addr, &gpc_dummy_register_map[addr]));
	}

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpc_handle_byte(addr | GP_MODE_WRITE));
		fail_unless(0 == gpc_handle_byte(data & 0xFF));
		fail_unless(0 == gpc_handle_byte(data >> 8));
		fail_unless(1 == gpc_dummy_register_changed);
		fail_unless(addr == gpc_dummy_register_changed_addr);
		fail_unless((void *)1 == gpc_dummy_register_changed_data);
		fail_unless(data == gpc_dummy_register_map[addr]);

		gpc_dummy_register_changed = 0;
		gpc_dummy_register_changed_addr = 0;
		gpc_dummy_register_changed_data = 0;
	}

	//for(addr=0; addr<32; addr++){
	//	fail_unless(1 == gpc_handle_byte(addr | GP_MODE_WRITE | GP_MODE_STRING));
	//	fail_unless(0 == gpc_dummy_register_changed);
	//	fail_unless(0 == gpc_dummy_register_changed_addr);
	//	fail_unless((void *)0 == gpc_dummy_register_changed_data);
	//	fail_unless(-1 == gpc_pickup_byte());
	//
	//	gpc_dummy_register_changed = 0;
	//	gpc_dummy_register_changed_addr = 0;
	//	gpc_dummy_register_changed_data = 0;
	//}
}
END_TEST

START_TEST(test_gprotc_read_cont)
{
	uint16_t addr = 0;

	for(addr=0; addr<32; addr++){
		fail_unless(0 == gpc_setup_reg(addr, &gpc_dummy_register_map[addr]));
	}

	for(addr=0; addr<32; addr++){
		gpc_dummy_register_map[addr] = addr | ((addr+1) << 8);
	}

	for(addr=0; addr<256; addr++){
		fail_unless(1 == gpc_register_touched(addr));
		fail_unless(0 == gpc_dummy_register_changed);
		fail_unless(0 == gpc_dummy_register_changed_addr);
		fail_unless((void *)0 == gpc_dummy_register_changed_data);
	}

	for(addr=0; addr<31; addr++){
		fail_unless(0 == gpc_handle_byte(addr | GP_MODE_READ | GP_MODE_CONT));
		fail_unless(0 == gpc_register_touched(addr));
		fail_unless(addr == gpc_pickup_byte());
		fail_unless(addr == gpc_pickup_byte());
		fail_unless(addr+1 == gpc_pickup_byte());
		fail_unless(-1 == gpc_pickup_byte());
		fail_unless(0 == gpc_handle_byte(addr | GP_MODE_READ | GP_MODE_CONT));
		fail_unless(1 == gpc_register_touched(addr));
		fail_unless(0 == gpc_dummy_register_changed);
		fail_unless(0 == gpc_dummy_register_changed_addr);
		fail_unless((void *)0 == gpc_dummy_register_changed_data);
	}
}
END_TEST

START_TEST(test_gprotc_send_short_string)
{
	int i;
	char *string = "Hello World!";
	int string_size = strlen(string);

	fail_unless(string_size == gpc_send_string(string, string_size));

	fail_unless((GP_MODE_STRING | string_size) == gpc_pickup_byte());

	for (i=0; i<string_size; i++) {
		fail_unless(string[i] == gpc_pickup_byte());
	}

	fail_unless(-1 == gpc_pickup_byte());
}
END_TEST

START_TEST(test_gprotc_send_long_string)
{
	int i,j;
	int string_size = (('z' - 'a') + 1) * ((('9' - '0') + 1) + 1);
	char string[string_size];

	/* Generate string. */
	{
		int gi, gj, gk = 0;

		for(gi=0; gi<(('z' - 'a') + 1); gi++) {
			string[gk++] = 'a'+gi;
			for(gj=0; gj<(('9' - '0') + 1); gj++){
				string[gk++] = '0'+gj;
			}
		}

		string[gk] = '\0';

	}

	/* Run test. */
	fail_unless(string_size == gpc_send_string(string, string_size));

	for (j=0; j<(string_size / GP_STR_PAK_MAX_LEN); j++) {
		fail_unless((GP_MODE_STRING | GP_STR_PAK_MAX_LEN) == gpc_pickup_byte());

		for (i=0; i<GP_STR_PAK_MAX_LEN; i++) {
			fail_unless(string[i + (j * GP_STR_PAK_MAX_LEN)] == gpc_pickup_byte());
		}
	}

	fail_unless((GP_MODE_STRING | (string_size % GP_STR_PAK_MAX_LEN)) == gpc_pickup_byte());


	for (i=0; i<(string_size % GP_STR_PAK_MAX_LEN); i++) {
		fail_unless(string[i + (j * GP_STR_PAK_MAX_LEN)] == gpc_pickup_byte());
	}

	fail_unless(-1 == gpc_pickup_byte());
}
END_TEST

START_TEST(test_gprotc_send_version)
{
	int32_t ch;
	int32_t to_read;
	int i, j;
	char string[3][128];

        fail_unless(0 == gpc_handle_byte(GP_MODE_STRING));

	/* we are expecting to receive three strings */
	for (i = 0; i<3; i++) {
		ch = gpc_pickup_byte();
		to_read = (ch & ~GP_MODE_STRING);
		fail_unless(ch == (GP_MODE_STRING | to_read));

		for (j=0; j<to_read; j++) {
			ch = gpc_pickup_byte();
			fail_if(-1 == ch);
			string[i][j] = ch;
			//putchar(ch);
		}
	}

	fail_unless(0 == regmatch("^libgovernor [[:digit:]]+\\.[[:digit:]]+-[[:alnum:]]{8}(-dirty)?, build [[:digit:]]{8}$", string[0]));
	fail_unless(0 == regmatch("^Copyright \\(C\\) 2010-20[[:digit:]]{2} Piotr Esden-Tempski <piotr@esden.net>$", string[1]));
	fail_unless(0 == regmatch("^License GPLv3\\+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>$", string[2]));

	fail_unless((void *)1 == gpc_dummy_get_version_data);
	fail_unless(1 == gpc_dummy_get_version_triggered);
}
END_TEST

Suite *make_lg_gprotc_suite(void)
{
	Suite *s;
	TCase *tc;

	s= suite_create("Governor protocol for client");

	tc = tcase_create("Basic functions");
	suite_add_tcase(s, tc);
	tcase_add_checked_fixture(tc, init_gprotc_tc, clean_gprotc_tc);
	tcase_add_test(tc, test_gprotc_send_reg);
	tcase_add_test(tc, test_gprotc_handle_byte_read);
	tcase_add_test(tc, test_gprotc_handle_byte_write);
	tcase_add_test(tc, test_gprotc_read_cont);
	tcase_add_test(tc, test_gprotc_send_short_string);
	tcase_add_test(tc, test_gprotc_send_long_string);
	tcase_add_test(tc, test_gprotc_send_version);

	return s;
}


include/lg/ring.h
/*
 * libgovernor - Open-BLDC configuration and debug protocol library
 * Copyright (C) 2010 by Piotr Esden-Tempski <piotr@esden.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef RING_H
#define RING_H

#include <stdint.h>

typedef int32_t ring_size_t;

struct ring {
	uint8_t *data;
	ring_size_t size;
	uint32_t begin;
	uint32_t end;
};

#define RING_SIZE(RING) ((RING)->size - 1)
#define RING_DATA(RING) (RING)->data

void ring_init(struct ring *ring, uint8_t * buf, ring_size_t size);
int32_t ring_write_ch(struct ring *ring, uint8_t ch);
int32_t ring_write(struct ring *ring, uint8_t * data, ring_size_t size);
int32_t ring_safe_write_ch(struct ring *ring, uint8_t ch);
int32_t ring_safe_write(struct ring *ring, uint8_t * data, ring_size_t size);
int32_t ring_read_ch(struct ring *ring, uint8_t * ch);
int32_t ring_read(struct ring *ring, uint8_t * data, ring_size_t size);

#endif /* RING_H */


src/gprotc.c
/*
 * libgovernor - Open-BLDC configuration and debug protocol library
 * Copyright (C) 2009-2010 by Piotr Esden-Tempski <piotr@esden.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//#define GPC_DEBUG

#ifdef GPC_DEBUG
#include <stdio.h>
#define DEBUG(STR, ARGS...) printf(STR , ## ARGS)
#else
#define DEBUG(STR, ARGS...)
#endif

#include "config.h"

#include <stdint.h>
#include <string.h>

#include "lg/ring.h"
#include "lg/gpdef.h"

#include "lg/gprotc.h"

struct gpc_hooks {
	gp_simple_hook_t trigger_output;
	void *trigger_output_data;
	gp_with_addr_hook_t register_changed;
	void *register_changed_data;
	gp_simple_hook_t get_version;
	void *get_version_data;
} gpc_hooks;

volatile uint16_t *gpc_register_map[32];

#define GPC_OUTPUT_BUFFER_SIZE 1024

struct ring gpc_output_ring;
uint8_t gpc_output_buffer[GPC_OUTPUT_BUFFER_SIZE];

enum gpc_states {
	GPCS_IDLE,
	GPCS_DATA_LSB,
	GPCS_DATA_MSB
};
enum gpc_states gpc_state = GPCS_IDLE;
uint16_t gpc_addr;
uint16_t gpc_data;

uint32_t gpc_monitor_map;

#ifndef PACKAGE_STRING
#define PACKAGE_STRING "null"
#endif

#ifndef VERSION_SUFFIX
#define VERSION_SUFFIX "null"
#endif

#ifndef BUILDDATE
#define BUILDDATE "null"
#endif

#define GPC_VERSION PACKAGE_STRING VERSION_SUFFIX ", build " BUILDDATE "\n"
#define GPC_COPYRIGHT COPYRIGHT "\n"
#define GPC_LICENSE LICENSE "\n"

int gpc_init(gp_simple_hook_t trigger_output, void *trigger_output_data,
	     gp_with_addr_hook_t register_changed, void *register_changed_data)
{
	int i;

	gpc_state = GPCS_IDLE;

	gpc_hooks.trigger_output = trigger_output;
	gpc_hooks.trigger_output_data = trigger_output_data;
	gpc_hooks.register_changed = register_changed;
	gpc_hooks.register_changed_data = register_changed_data;
	gpc_hooks.get_version = 0;
	gpc_hooks.get_version_data = 0;

	for (i = 0; i < 32; i++)
		gpc_register_map[i] = 0;

	gpc_monitor_map = 0;

	ring_init(&gpc_output_ring, gpc_output_buffer, GPC_OUTPUT_BUFFER_SIZE);

	return 0;
}

int gpc_set_get_version_callback(gp_simple_hook_t get_version, void *get_version_data)
{

	gpc_hooks.get_version = get_version;
	gpc_hooks.get_version_data = get_version_data;

	return 0;
}

int gpc_setup_reg(uint8_t addr, volatile uint16_t * reg)
{
	if (addr > 31)
		return 1;

	DEBUG("Setting up register %02X with %p\n", addr, reg);

	gpc_register_map[addr] = reg;

	return 0;
}

int32_t gpc_pickup_byte(void)
{
	return ring_read_ch(&gpc_output_ring, 0);
}

int gpc_send_reg(uint8_t addr)
{
	uint8_t dat[3];

	if ((addr > 31) | !gpc_register_map[addr])
		return 1;

	dat[0] = addr;
	dat[1] = (*gpc_register_map[addr]) & 0xFF;
	dat[2] = (*gpc_register_map[addr]) >> 8;

	DEBUG("sending reg %02X with content %04X\n", addr,
	      *gpc_register_map[addr]);

	if (0 <= ring_write(&gpc_output_ring, dat, 3)) {
		if (gpc_hooks.trigger_output)
			gpc_hooks.trigger_output(gpc_hooks.trigger_output_data);
		return 0;
	}

	return 1;
}

int gpc_send_string(char *string, int len)
{
	int i;

	for (i=0; i<(len / GP_STR_PAK_MAX_LEN); i++) {
		/* Send out the start byte for a string */
		if (0 > ring_safe_write_ch(&gpc_output_ring, GP_MODE_STRING |
						GP_STR_PAK_MAX_LEN)) {
			return -1;
		}

		/* Send packet contents */
		if (0 > ring_safe_write(&gpc_output_ring, (uint8_t *)(string + (i * GP_STR_PAK_MAX_LEN)), GP_STR_PAK_MAX_LEN)) {
			return -1;
		}
	}

	if (0 > ring_safe_write_ch(&gpc_output_ring, GP_MODE_STRING | (len % GP_STR_PAK_MAX_LEN))) {
		return -1;
	}

	if (0 > ring_safe_write(&gpc_output_ring, (uint8_t *)(string + (i * GP_STR_PAK_MAX_LEN)), (len % GP_STR_PAK_MAX_LEN))) {
		return -1;
	}

	if (gpc_hooks.trigger_output)
		gpc_hooks.trigger_output(gpc_hooks.trigger_output_data);

	return len;
}

int gpc_handle_byte(uint8_t byte)
{
	DEBUG("got byte %04X ", byte);

	switch (gpc_state) {
	case GPCS_IDLE:
		if (byte & GP_MODE_STRING) {
			if (byte == GP_MODE_STRING) {
				gpc_send_string(GPC_VERSION, sizeof(GPC_VERSION) - 1);
				gpc_send_string(GPC_COPYRIGHT, sizeof(GPC_COPYRIGHT) - 1);
				gpc_send_string(GPC_LICENSE, sizeof(GPC_LICENSE) - 1);
				if (gpc_hooks.get_version)
					gpc_hooks.get_version(gpc_hooks.get_version_data);
				return 0;
			}
			DEBUG("not handled\n");
			return 1;
		}

		if ((byte & GP_MODE_MASK) == (GP_MODE_WRITE)) {
			DEBUG("write ");
			gpc_addr = byte & GP_ADDR_MASK;
			gpc_state = GPCS_DATA_LSB;
		} else if ((byte & GP_MODE_MASK) ==
			   (GP_MODE_READ | GP_MODE_PEEK)) {
			DEBUG("read ");
			gpc_send_reg(byte & GP_ADDR_MASK);
		} else if ((byte & GP_MODE_MASK) ==
			   (GP_MODE_READ | GP_MODE_CONT)) {
			DEBUG("read cont ");
			gpc_monitor_map ^= 1 << (byte & GP_ADDR_MASK);
		} else {
			DEBUG("unimplemented\n");
			return 1;
		}
		break;
	case GPCS_DATA_LSB:
		DEBUG("lsb ");
		gpc_data = byte;
		gpc_state = GPCS_DATA_MSB;
		break;
	case GPCS_DATA_MSB:
		DEBUG("msb ");
		gpc_data |= byte << 8;
		gpc_state = GPCS_IDLE;

		if (!gpc_register_map[gpc_addr]) {
			DEBUG("addr %02X with pointer %p not set up\n",
			      gpc_addr, gpc_register_map[gpc_addr]);
			return 1;
		}

		*gpc_register_map[gpc_addr] = gpc_data;
		if (gpc_hooks.register_changed)
			gpc_hooks.register_changed(gpc_hooks.
						   register_changed_data,
						   gpc_addr);

		break;
	default:
		return 1;
	}

	DEBUG("\n");
	return 0;
}

int gpc_register_touched(uint8_t addr)
{
	if (addr > 31)
		return 1;

	DEBUG("touched_register %02X, search mask %08X and register map %08X ",
	      addr, (1 << addr), gpc_monitor_map);
	if (gpc_monitor_map & (1 << addr)) {
		DEBUG("sending\n");
		gpc_send_reg(addr);
	} else {
		DEBUG("not sending\n");
		return 1;
	}

	return 0;
}
